File: backend/src/main/java/com/forensic/service/TeamService.java
package com.forensic.service;

import com.forensic.dto.TeamDto;
import com.forensic.entity.CaseAssignment;
import com.forensic.repository.CaseAssignmentRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class TeamService {

    private final CaseAssignmentRepository caseAssignmentRepository;

    /**
     * Get all teams (group active assignments by caseId)
     */
    public List<TeamDto> getAllTeams() {

        List<CaseAssignment> activeAssignments =
                caseAssignmentRepository.findAllActive();

        Map<Long, List<Long>> teamsMap =
                activeAssignments.stream()
                        .collect(Collectors.groupingBy(
                                CaseAssignment::getCaseId,
                                Collectors.mapping(
                                        CaseAssignment::getUserId,
                                        Collectors.toList()
                                )
                        ));

        return teamsMap.entrySet()
                .stream()
                .map(e -> new TeamDto(e.getKey(), e.getValue()))
                .toList();
    }

    /**
     * Get one team by caseId
     */
    public TeamDto getTeamByCaseId(Long caseId) {

        List<Long> userIds = caseAssignmentRepository
                .findActiveByCaseId(caseId)
                .stream()
                .map(CaseAssignment::getUserId)
                .toList();

        return new TeamDto(caseId, userIds);
    }

    /**
     * Add member to team
     */
    @Transactional
    public void addMemberToTeam(Long caseId, Long userId) {

        Optional<CaseAssignment> existing =
                caseAssignmentRepository.findByCaseIdAndUserId(caseId, userId);

        if (existing.isPresent()) {
            CaseAssignment assignment = existing.get();

            if (assignment.getRemovedAt() == null) {
                return; // already active
            }

            assignment.setRemovedAt(null);
            assignment.setAssignedAt(LocalDateTime.now());
            caseAssignmentRepository.save(assignment);

        } else {
            CaseAssignment newAssignment = new CaseAssignment();
            newAssignment.setCaseId(caseId);
            newAssignment.setUserId(userId);
            newAssignment.setAssignedAt(LocalDateTime.now());
            newAssignment.setRemovedAt(null);

            caseAssignmentRepository.save(newAssignment);
        }
    }

    /**
     * Remove member from team (soft delete)
     */
    @Transactional
    public void removeMemberFromTeam(Long caseId, Long userId) {
        caseAssignmentRepository.revokeUserFromCaseTeam(caseId, userId);
    }
}

File: frontend/src/services/teamService.js
import axios from "axios";

const API = "http://localhost:8443/api/team";

export const getAllTeams = () =>
  axios.get(API);

export const getTeamByCaseId = (caseId) =>
  axios.get(`${API}/${caseId}`);

export const addMemberToTeam = (caseId, userId) =>
  axios.post(`${API}/${caseId}/members/${userId}`);

export const removeMemberFromTeam = (caseId, userId) =>
  axios.delete(`${API}/${caseId}/members/${userId}`);

// TeamService.js
// PURE STATIC JSON — NO API, NO MUTATION

// Static team data matching your 6 cases
// const teamsData = [
//   { caseId: 1, userIds: [101, 102] },
//   { caseId: 2, userIds: [103] },
//   { caseId: 3, userIds: [] },
//   { caseId: 4, userIds: [101] },
//   { caseId: 5, userIds: [104, 105] },
//   { caseId: 6, userIds: [] }
// ];

// // ---------------- DUMMY GET ALL ----------------
// export const getAllTeams = () => {
//   return new Promise((resolve) => {
//     resolve({ data: teamsData });
//   });
// };